test/iidm/extensions/SlackTerminalTest.cpp

```diff
@@ -0,0 +1,240 @@
+
+    for (VoltageLevel& vl : network.getVoltageLevels()) {
+        auto slackTerminal = vl.findExtension<SlackTerminal>();
+        if (slackTerminal) {
+            BOOST_CHECK_EQUAL(busBusBreakerId, slackTerminal.get().getTerminal().get().getBusBreakerView().getBus().get().getId());
+            BOOST_CHECK_EQUAL("VL_0", slackTerminal.get().getTerminal().get().getBusView().getBus().get().getId());
+        }
+    }
+}
+
+BOOST_AUTO_TEST_CASE(variantsCloneTest) {
+    std::string variant1 = "variant1";
+    std::string variant2 = "variant2";
+    std::string variant3 = "variant3";
+
+    // Creates the extension
+    Network network = powsybl::network::EurostagFactory::createTutorial1Network();
+    VoltageLevel& vl = network.getVoltageLevel("VLLOAD");
+    SlackTerminal::attach(network.getBusBreakerView().getBus("NLOAD"));
+    auto& slackTerminal = vl.getExtension<SlackTerminal>();
+    Terminal& t0 = slackTerminal.getTerminal();
+
+    // Testing variant cloning
+    VariantManager& variantManager = network.getVariantManager();
+    variantManager.cloneVariant(VariantManager::getInitialVariantId(), variant1);
+    variantManager.cloneVariant(variant1, variant2);
+    variantManager.setWorkingVariant(variant1);
+    BOOST_CHECK(stdcxx::areSame(t0, slackTerminal.getTerminal().get()));
+
+    // Removes a variant then adds another variant to test variant recycling (hence calling allocateVariantArrayElement)
+    variantManager.removeVariant(variant1);
+    std::initializer_list<std::string> targetVariantIds = {variant1, variant3};
+    variantManager.cloneVariant(VariantManager::getInitialVariantId(), targetVariantIds);
+    variantManager.setWorkingVariant(variant1);
+    BOOST_CHECK(stdcxx::areSame(t0, slackTerminal.getTerminal().get()));
+    variantManager.setWorkingVariant(variant3);
+    BOOST_CHECK(stdcxx::areSame(t0, slackTerminal.getTerminal().get()));
+
+    // Test removing current variant
+    variantManager.removeVariant(variant3);
+    POWSYBL_ASSERT_THROW(slackTerminal.getTerminal(), PowsyblException, "Variant index not set");
+}
+
+BOOST_AUTO_TEST_CASE(vlErrorTest) {
+    Network network = powsybl::network::EurostagFactory::createTutorial1Network();
+    VoltageLevel& vl = network.getVoltageLevel("VLHV1");
+
+    // Adding a terminal in the wrong voltage level
+    Terminal& wrongTerminal = *network.getBusBreakerView().getBus("NLOAD").get().getConnectedTerminals().begin();
+    SlackTerminalAdder slackTerminalAdder = vl.newExtension<SlackTerminalAdder>().withTerminal(wrongTerminal);
+    POWSYBL_ASSERT_THROW(slackTerminalAdder.add(), PowsyblException, "Terminal given is not in the right VoltageLevel (VLLOAD instead of VLHV1)");
+
+    // First adding a terminal in the right voltage level...
+    Terminal& terminal = *boost::begin(network.getBusBreakerView().getBus("NHV1").get().getConnectedTerminals());
+    slackTerminalAdder.withTerminal(terminal).add();
+    SlackTerminal slackTerminal = vl.getExtension<SlackTerminal>();
```

Comment:
Extension should not be copiable?

Reply:
This extension is copiable like every other extension. Do you mean by deleting copy constructor and affectation operator ? If yes, to be done in a separated PR (IMHO).

Reply:
But this is a mistake: extensions should not be copiable because there is a double-link between the extendable and the extension. By definition, an extendable can carry only a single instance of each extensions.

Agree to do it in a separate PR: could you please prepare an issue and assign it in the current project. From my point of view, this is a critical design issue

Reply:
Ok, done : https://github.com/powsybl/powsybl-iidm4cpp/issues/203